client/server/bundler/node-moduler.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * External Dependencies
+ */
+const fs = require( 'fs' ); // eslint-disable-line  import/no-nodejs-modules
+const { builtinModules } = require( 'module' );
+const rcopy = require( 'recursive-copy' );
+const mkdirp = require( 'mkdirp' );
+
+function getModule( request ) {
+	const parts = request.split( '/' );
+	if ( parts[ 0 ].startsWith( '@' ) ) {
+		return parts[ 0 ] + '/' + parts[ 1 ];
+	}
+	return parts[ 0 ];
+}
+
+module.exports = class NodeModuler {
+	apply( compiler ) {
+		compiler.hooks.afterEmit.tapAsync( 'NodeModuler', ( compilation ) => {
+			const externalModules = [];
+			for ( const module of compilation.modules ) {
+				if ( module.external ) {
+					const requestModule = getModule( module.userRequest );
+
+					if ( builtinModules.includes( requestModule ) ) {
+						// native Node.js module, not in node_modules
+						continue;
+					}
+
+					if ( requestModule.startsWith( './' ) || requestModule.startsWith( '../' ) ) {
+						// loading local file by relative path, not in node_modules
+						continue;
+					}
+
+					if ( ! externalModules.includes( requestModule ) ) {
+						console.log( 'adding to node_modules because required by bundle:', requestModule );
+						externalModules.push( requestModule );
+					}
+				}
+			}
+
+			const visitedFolders = new Set();
+
+			function collectDependencies( mod, contexts ) {
+				console.log( 'collecting', mod, contexts );
+				const modFolder = contexts[ 0 ] + '/' + mod;
+
+				// read module's package.json
+				const pkg = JSON.parse( fs.readFileSync( modFolder + '/package.json' ) );
+
+				// bail out if package has no dependencies
+				if ( ! pkg.dependencies ) {
+					console.log( 'Package ' + mod + ' has no dependencies, skipping' );
+					return;
+				}
+
+				// unshift the module's nested node_modules to contexts
+				contexts = [ modFolder + '/node_modules', ...contexts ];
+
+				// iterate its dependencies
+				for ( const dep of Object.keys( pkg.dependencies ) ) {
```

Comment:
Do we need to consider `peerDependencies` and `optionalDependencies`?

Reply:
I'd say we don't. `peerDependencies` don't trigger installation -- the depedency is supposed to be installed by someone else. Although I read some news that the very latest NPM will start installing them, too, in certain cases.

`optionalDependencies` are very rare, and optional. But it's possible that a missing optional dependency will disable some feature that was until now enabled, due to a conditional `require` succeeding. Worth auditing. It's not a problem for the plugin to start checking for them, too.

Reply:
My point is that if a module declares `peerDependencies` or `optionalDependencies`, and they are installed, we should copy them as well because chances are the module are importing them.